library/src/main/java/org/apache/apex/malhar/lib/fs/s3/S3OutputModule.java

```diff
@@ -0,0 +1,325 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.apex.malhar.lib.fs.s3;
+
+import javax.validation.constraints.Min;
+import javax.validation.constraints.NotNull;
+
+import org.apache.hadoop.classification.InterfaceStability;
+import org.apache.hadoop.conf.Configuration;
+
+import com.datatorrent.api.Context;
+import com.datatorrent.api.DAG;
+import com.datatorrent.api.Module;
+import com.datatorrent.common.partitioner.StatelessPartitioner;
+import com.datatorrent.lib.io.block.AbstractBlockReader;
+import com.datatorrent.lib.io.block.BlockMetadata;
+import com.datatorrent.lib.io.fs.AbstractFileSplitter;
+import com.datatorrent.netlet.util.Slice;
+
+import static com.datatorrent.api.Context.OperatorContext.TIMEOUT_WINDOW_COUNT;
+
+/**
+ * S3OutputModule can be used to upload the files/directory into S3. This module supports
+ * parallel uploads of multiple blocks of the same file and merge those blocks in sequence.
+ *
+ * Below operators are wrapped into single component using Module API
+ *  - S3InitiateFileUpload
+ *  - S3BlockUpload
+ *  - S3FileMerger
+ *
+ * Initial BenchMark Results
+ * -------------------------
+ * The Module writes 18 MB/s to S3 using multi part upload feature with the following configuration
+ *
+ * File Size = 1 GB
+ * Partition count of S3BlockUpload = 6
+ * Partition count of S3FileMerger = 1
+ * Container memory size of this module as follows:
+ *          S3InitiateFileUpload  = 1 GB
+ *          S3BlockUpload         = 2.5 GB
+ *          S3FileMerger          = 2 GB
+ *
+ *
+ *  @displayName S3 Output Module
+ *  @tags S3, Output
+ */
+@InterfaceStability.Evolving
+public class S3OutputModule implements Module
+{
+  /**
+   * AWS access key
+   */
+  @NotNull
+  private String accessKey;
```

Comment:
Created JIRA for this task: https://issues.apache.org/jira/browse/APEXMALHAR-2360